beacon_node/beacon_chain/src/blob_verification.rs

```diff
@@ -236,25 +252,15 @@ pub fn validate_blob_sidecar_for_gossip<T: BeaconChainTypes>(
         return Err(BlobError::ProposerSignatureInvalid);
     }
 
-    // TODO(pawan): kzg validations.
-
-    // TODO(pawan): Check if other blobs for the same proposer index and blob index have been
-    // received and drop if required.
-
-    // Verify if the corresponding block for this blob has been received.
-    // Note: this should be the last gossip check so that we can forward the blob
-    // over the gossip network even if we haven't received the corresponding block yet
-    // as all other validations have passed.
-    let block_opt = chain
-        .canonical_head
-        .fork_choice_read_lock()
-        .get_block(&block_root)
-        .or_else(|| chain.early_attester_cache.get_proto_block(block_root)); // TODO(pawan): should we be checking this cache?
-
-    // TODO(pawan): this may be redundant with the new `AvailabilityProcessingStatus::PendingBlock variant`
-    if block_opt.is_none() {
-        return Err(BlobError::UnknownHeadBlock {
-            beacon_block_root: block_root,
+    // Verify that this is the first blob sidecar received for the (sidecar.block_root, sidecar.index) tuple
```

Comment:
We need this check after sig verification to ensure that it's a valid blob signed by the proposer of that slot.

Agree that we need a separate seen cache. Working on adding one now.

Reply:
Just adding it for my notes here.

I think this `seen_cache` should be strictly gossip and not contain any blobs that we receive over rpc mainly because we do not receive proposer signed blobs over rpc. We shouldn't add anything in there which does not have any DoS protection imo. There could be an attack where:
1. attacker might send us an attestation to an unknown block hash
2. we start a a block and blobs by root request for the unknown hash
3. the attacker sends the valid block but sends us bogus blobs over rpc

If the bogus blobs get into the seen cache before the valid blob that is sent over gossip, we could end up ignoring the valid blobs.